src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs

```diff
@@ -358,6 +363,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
                 res.SetHandle(acceptedFd);
                 return res;
             }
+
+            internal unsafe bool TryUnblockSocket()
+            {
+                // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
+                // may block indefinitly. This is a best effort attempt to not get blocked and make those operations return.
+                // We need to ensure we keep the expected TCP behavior that is observed by the socket peer (FIN vs RST close).
+                // What we do here isn't specified by POSIX and doesn't work on all OSes.
+                // On Linux this works well.
+                // On OSX, TCP connections will be closed with a FIN close instead of an abortive RST close.
+                // And, pending TCP connect operations and UDP receive are not abortable.
+
+                // Don't touch sockets which don't have the CLOEXEC flag set. These may be shared
+                // with other processes and we want to avoid disconnecting them.
+                int fdFlags = Interop.Sys.Fcntl.GetFD(this);
+                if (fdFlags == 0)
+                {
+                    return false;
+                }
+
+                int type = 0;
+                int optLen = sizeof(int);
+                Interop.Error err = Interop.Sys.GetSockOpt(this, SocketOptionLevel.Socket, SocketOptionName.Type, (byte*)&type, &optLen);
+                Debug.Assert(err == Interop.Error.SUCCESS);
```

Comment:
In what situation might this fail?  If it does fail, what will our follow-up action be... remove the assert?  We generally don't assert on the results of P/Invokes like this... what makes this one ok?

Reply:
I'll remove this.